arrow2_convert_derive/src/derive_enum.rs

```diff
@@ -227,6 +226,7 @@ pub fn expand_serialize(input: DeriveEnum) -> TokenStream {
         let first_name = &variant_names[0];
         quote! {
             self.types.push(0);
+            self.offsets.push((self.#first_name.len()) as i32);
```

Comment:
This appears to have been a more general bug with pushing None values to dense unions... but `Option<>` of enum isn't a very common pattern so I'm not shocked we hadn't hit it prior to sticking a dense union inside a sparse union.